Observaciones de estructura

rogs-lv opened the pull request:
1. Agrega meaningful
2. Correccion de complejidad para SumArray
3. Modificaciones a PR
4. Código limpio

rogs-lv committed:
Agrega observaciones de estructura

1. Agrega meaningful
2. Correccion de complejidad para SumArray
3. Modificaciones a PR
4. Código limpio

jorgetinoco commented on Algorithms.Test/ShuffleTheArrayTest.cs:
```diff
@@ -0,0 +1,69 @@
+        [Fact]
+        public void TestCase1() {
+            //Input
+            int[] numsArray     = new int[] { 2, 5, 1, 3, 4, 7 };
```
adding the spaces between the assignation is generaly not a good practice, since adding a new line requires adjustment on each line

rogs-lv replied:
Got it!. I read once, that my code should be easy to read, like a book. I assumed if I used tabs or extra space. It would achieve the goal. I think ... i didn't understand well hehe

jorgetinoco replied:
That's true, it should be easy to read as a book. One part of that is to use spaces, but in a different way for example:
What's easier to read, this:
```js
let y = 10;
let x = 1;
if(y==='division')
return x/y;
else if(y==='mult')
return x*y;
return x+y;
```

or this:
```js
let operator = 10;
let myNumber = 1;

if (operation === 'division')
   return myNumber / operator;
else if (operation === 'mult')
  return myNumber * operator;

return myNumber + operator;
```

If you see, I changed 2 things, one was spaces/tabs and the second was variable naming. This "easy to read" heavily applies to variable naming and "cleaningless" of the code, reading x + y doesn't give you any context or what x or y means, while a different variable name can give you more information so you can keep reading the code without reading documentation, reading external functions or any other thing


So yes, that's true adding spaces helps, but you should think also in maintainability, for example

```js
let x          = true
let y          = false
let z          = 1 
```

and what happen if you want to add a variable called `superLongName`:
```js
let x          = true
let y          = false
let z          = 1 
let superLongName = true
```

Then you have either of 2 problems, they are not aligned anymore, so not easy to read, or you need to update 3 lines instead of just adding your variable.

rogs-lv replied:
Great!! I definitely didn't get it right. Now i see the problem 👏

jorgetinoco commented on Algorithms/RunningSumArray.cs:
```diff
@@ -0,0 +1,56 @@
+            {
+                if (i > 0)
+                    nums[i] = nums[i - 1] + nums[i];
+                else
```
there's a way you can remove this if/else statement.

rogs-lv replied:
Yes, There's a way. Actually first element of array never is affected with new value. Therefore loop for can start in 1.

jorgetinoco commented on Algorithms.Test/ShuffleTheArrayTest.cs:
```diff
@@ -0,0 +1,69 @@
+            Assert.Equal(outputArray, result);
+        }
+        [Fact]
+        public void TestCase4() {
```
Imagine this is a real project. TestCase4 doesn't tell me what this test actually do. You can check unit tests best practices here:
https://docs.microsoft.com/en-us/dotnet/core/testing/unit-testing-best-practices#naming-your-tests

rogs-lv replied:
I check the information 💯

jorgetinoco commented on Algorithms.Test/ShuffleTheArrayTest.cs:
```diff
@@ -0,0 +1,69 @@
+        public void TestCase5() {
+            //Input
+            int[] numsArray     = new int[] { 361, 681, 821, 914, 611, 137, 702, 8, 956, 598, 65, 921, 678, 476, 522, 961, 611, 752, 121, 756, 439, 640, 890, 171, 462, 942, 141, 832, 317, 251, 27, 824, 71, 781, 968, 177, 763, 43, 257, 430, 236, 612, 619, 3, 390, 900, 169, 427, 112, 628, 434, 891, 636, 68, 401, 720, 790, 651, 453, 668, 996, 870, 691, 705, 571, 241, 850, 897, 508, 345, 199, 568, 49 ...
+            int elements        = 420;
```
As you can see, all your tests are the happy path. 
What happens if I sent `elements = -1`, what happens if I sent `nums Array = new int[0]` and then `elements = 10`... etc etc. That's the meaning of edge cases

rogs-lv replied:
Got it. I'll change my unit tests

jorgetinoco commented on Algorithms/Program.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace Algorithms
+{
+    public class Program
```
probably you don't need this class

rogs-lv replied:
Got it!.

jorgetinoco commented on Algorithms/RunningSumArray.cs:
```diff
@@ -0,0 +1,56 @@
+        }
+        /// <summary>
+        /// Big O 
+        /// Complejidad: O(N2) - Se tienen dos ciclos recorriendo el mismo arreglo, y el segundo lo recorre
```
👍

rogs-lv replied:
👍 Cool!

jorgetinoco commented on README.md:
```diff
@@ -5,6 +5,16 @@ Leetcode exercise for training. A solution with multiple console applications.
 
 Run `dotnet build` to build the project.
 
+## Test
```
A good practice in the README is to mention the minimum version you need to run the project. For example:

## Build

To build this project you need at least NetCore 3.1.
Then you can build the project with

```bash
dotnet build
```

rogs-lv replied:
Perfect! :)

jorgetinoco commented on Algorithms/RunningSumArray.cs:
```diff
@@ -0,0 +1,56 @@
+        /// Complejidad: O(n) - El arreglo solo se recorre una unica vez y se le resta la posición 
+        /// anterior que ya tiene el ultimo valor acumulado
+        /// </summary>
+        /// <param name="nums">Array de enteros</param>
```
Good coding practices:
If the parameter is called `nums` it can be anything, and then you end up with documentation like this one `Array de enteros`, here are two things:
- Imagine your code is international, it is always a good practice to do your code in a "neutral" language as english so everyone can understand it
- What could be a better variable naming? `numArray` for example and that could give you more information without reading the documentation.


Apply the same to other parts of the code, for example in shuffleArray, what's `l`, `j` ?

rogs-lv replied:
OK, So i'll change comments.

jorgetinoco reviewed (changes requested):
Hey @rogs-lv I did some comments.

Some of them may look "overkill" for this project but remember the reason behind this exercise, which is improve in all aspects, that's why I'm being so picky about clean code, tests etc.

Also feel free to add here the exercise you sent me, the palindrome solution

jorgetinoco commented:
@rogs-lv 
Also try to follow git/github good practices for this small project.
One example I have right now is about commits, it is easier to follow the history of the commits and review changes if commits are separated meaningfully. For example in this case I only see 1 commit `Agrega observaciones de estructura` which in my head I can only think: `which observations?` `which structure?`.
Eventhough you added the comment in the commit, it would be easier to see the following (example) commit structure:

```bash
fix variables to have meaningful names
fix SumArray, make it O(n)
Improve clean code
```
Inside each commit you can explain more about what it is about, but then if I want to review only how you changed SumArray to be O(n) I can go to that commit and see the changes.

https://medium.com/@nawarpianist/git-commit-best-practices-dab8d722de99

rogs-lv committed:
Aplica correcciones de mejores practicas

* Cambia nombres de metodos de pruebas unitarias como mejores practicas
* Aplica edge case para mejora pruebas unitarias
* Cambia descripción de complejidad
* Remueve validación inecesaria para RunningSum
* Corrije detalle de README para seccion de Build

jorgetinoco commented on Algorithms/ShuffleTheArray.cs:
```diff
@@ -0,0 +1,44 @@
+            int[] newArr = new int[lenArray];
+            bool flag = true;
+            int j = n;
+            int l = 0;
```
What about this? What's `l`?

jorgetinoco commented on Algorithms/ShuffleTheArray.cs:
```diff
@@ -0,0 +1,44 @@
+
+            int lenArray = numsArray.Length;
+            int[] newArr = new int[lenArray];
+            bool flag = true;
```
can you think in a better name than `flag`?

rogs-lv replied:
Sure, i think so.

jorgetinoco replied:
Just want to point this out. After thinking in better naming, you figure it out how to clean your code and ended up with a better version of the solution. That's what you need to do everytime :)

rogs-lv replied:
I understand, thanks for this advice 👌  👍 . In fact I always try, but sometimes I get stuck 😕

jorgetinoco commented on Algorithms/ShuffleTheArray.cs:
```diff
@@ -0,0 +1,44 @@
+                {
+                    newArr[i] = numsArray[l];
+                    l++;
+                    flag = false;
```
Can you think in a way to improve this `flag` code? In how many place you need to update the flag logic if for some reason it changes?

rogs-lv replied:
Yes, i can. I think is better if i change logic of method.

jorgetinoco commented on Algorithms.Test/RunningSumArrayTest.cs:
```diff
@@ -0,0 +1,47 @@
+            Assert.Empty(result);
+        }
+        [Fact]
+        public void RunningSum_ArrayWithOneElement_ReturnSameArray()
```
better tests!

jorgetinoco commented on Algorithms.Test/RunningSumArrayTest.cs:
```diff
@@ -0,0 +1,47 @@
+            _runningSumArray = new RunningSumArray();
+        }
+        [Fact]
+        public void RunningSum_SmallArray_ReturnRunningSumOfNumbers()
```
What about testing negative numbers inside the array?

jorgetinoco commented on Algorithms/RunningSumArray.cs:
```diff
@@ -0,0 +1,55 @@
+        /// </summary>
+        /// <param name="numsArray">Numbers array</param>
+        /// <returns>Return the running sum of numbers.</returns>
+        public int[] RunningSum(int[] numsArray)
```
what do you think about this code now? Comparing it to the first version

rogs-lv replied:
I think it's better, because by use the same array to overwrite value, i don't need more memory space. I also think it's cleaner to reduce the number of lines of code.

rogs-lv committed:
Change the logic method Shuffle

* Rename variables
* Improve method Shuffle with new logic

rogs-lv committed:
Add validation to method RunningSum

* Maximum and minimum value for array values.
* New unit tests to validate the maximum and minimum number for array.

rogs-lv reviewed (commented):
Applied

jorgetinoco commented on Algorithms.Test/RunningSumArrayTest.cs:
```diff
@@ -40,6 +40,21 @@ public void RunningSum_ArrayWithOneElement_ReturnSameArray()
+        }
+        [Fact]
+        public void RunningSum_ArrayWithNegativeNumber_ReturnSameArray() {
+            int[] numsArray = new int[] { -1000001, 2, 1 };
```
nice catch! 
But my question remains... what would happen if I send { -1, -2, -3, 5 }
(I know the code, so I know it would work, but there're missing tests about those cases)

jorgetinoco commented on Algorithms/ShuffleTheArray.cs:
```diff
@@ -7,38 +7,59 @@ namespace Algorithms
-                    j++;
-                    flag = true;
-                }
+                newArray[i + i] = numsArray[i];
```
🚀

rogs-lv committed:
Add validation test for numbers negatives

rogs-lv committed:
Add new excersice of Palindrome

* Add unit tests to excersice
* Add method to determine if number is palindrome

rogs-lv commented:
Add a new exercise with two solutions, the first solution had a little problem and was slower. But after thinking about what you told me about the name of the variables 🤔. I saw a new way to change the algorithm 🤓 . 
In this exercise i applied all the things we have been working on. 😃

jorgetinoco commented on Algorithms/PalindromeNumber.cs:
```diff
@@ -0,0 +1,59 @@
+            if (number < 0) return false; //Numbers negatives never will true
+            if (number <= 9 && number >= 0) return true; // Numbers with only digits always will true
+
+            int leng = number.ToString().Length;
```
length of what?

jorgetinoco commented on Algorithms/PalindromeNumber.cs:
```diff
@@ -0,0 +1,59 @@
+            if (number <= 9 && number >= 0) return true; // Numbers with only digits always will true
+
+            int leng = number.ToString().Length;
+            char[] arr = number.ToString().ToCharArray();
```
what is arr?

jorgetinoco commented on Algorithms/PalindromeNumber.cs:
```diff
@@ -0,0 +1,59 @@
+            char[] arr = number.ToString().ToCharArray();
+
+            int midpoint = (leng%2) == 1 ? (leng - 1) / 2 : leng / 2; // 1 is not even number || 0 is even number
+            for (int i = 0; i < midpoint; i++)
```
good one! This is the best way to check for palindromes. 
You will see different things like reverse, or things like that but this way is "better", and actually this part could be O(n/2) since you are not iterating the whole array, you are iterating half of it

jorgetinoco commented on Algorithms.Test/PalindromeNumberTest.cs:
```diff
@@ -0,0 +1,59 @@
+            Assert.True(result);
+        }
+        [Fact]
+        public void IsPalindrome_NumberNegative_ReturnFalse()
```
I'm guessing you used the one in leetcode for this exercise.
No need to do it, but how would you do it if for the problem didn't matter if it is negative or positive?
for example -515 should be true. What do you think you could do?

rogs-lv committed:
Add exercise ValidParentheses

rogs-lv committed:
Fixed variables name

rogs-lv committed:
Merge branch 'rama-ValidParentheses' into rama-ChangeStructure

rogs-lv committed:
Add exercise JewelsAndStones

jorgetinoco commented on Algorithms/ValidParentheses.cs:
```diff
@@ -0,0 +1,43 @@
+					return false;
+				i++;
+			}
+			return characters.Count == 0 ? true : false;
```
how could you improve this last line?

jorgetinoco commented on Algorithms/ValidParentheses.cs:
```diff
@@ -0,0 +1,43 @@
+		public bool IsValid(string stringCharacters) {
+			if (stringCharacters[0] == ')' || stringCharacters[0] == '}' || stringCharacters[0] == ']') return false;
+
+			Stack<char> characters = new Stack<char>();
```
nice

jorgetinoco commented on Algorithms/ValidParentheses.cs:
```diff
@@ -0,0 +1,43 @@
+			if (stringCharacters[0] == ')' || stringCharacters[0] == '}' || stringCharacters[0] == ']') return false;
+
+			Stack<char> characters = new Stack<char>();
+			Dictionary<char, char> charsOfClose = new Dictionary<char, char>();
```
nice! good way of improving the complexity :)

jorgetinoco commented on Algorithms/ValidParentheses.cs:
```diff
@@ -0,0 +1,43 @@
+
+			while (i < lengthString)
+			{
+				if (stringCharacters[i] == '(' || stringCharacters[i] == '{' || stringCharacters[i] == '[')
```
Check this documentation: https://github.com/ryanmcdermott/clean-code-javascript#encapsulate-conditionals
It is for JS, but the concepts can be applied in all languages (almost everything at least)
Check specific the one I shared `Encapsulate conditionals`

jorgetinoco commented on Algorithms/ValidParentheses.cs:
```diff
@@ -0,0 +1,43 @@
+		/// <param name="stringCharacters">string of characters</param>
+		/// <returns>true if all characters was closed, false if some characters wasn't closed</returns>
+		public bool IsValid(string stringCharacters) {
+			if (stringCharacters[0] == ')' || stringCharacters[0] == '}' || stringCharacters[0] == ']') return false;
```
what happens with an empty or null string?
You missed that unit test :)

jorgetinoco commented on Algorithms/JewelsAndStones.cs:
```diff
@@ -0,0 +1,41 @@
+            for (int i = 0; i < charsToSearch.Length; i++)
+            {
+                dictionaryJewels[charsToSearch[i]] = 0;
+            }
```
What could be a "break first" approach here?
What if you don't have any charsToSearch? do you still need to check the string?

jorgetinoco commented on Algorithms/JewelsAndStones.cs:
```diff
@@ -0,0 +1,41 @@
+                    dictionaryJewels[charsString[j]]++;
+            }
+
+            foreach (var i in dictionaryJewels)
```
is there a way you can remove this last foreach?

rogs-lv committed:
Add and fixed ValidParentheses

- Add unit test of string empty
- Apply best practice to encapsulate conditions
- Improve return of method

jorgetinoco commented on Algorithms/ValidParentheses.cs:
```diff
@@ -22,22 +22,38 @@ public class ValidParentheses
+		/// </summary>
+		/// <param name="stringCharacters">string of characters</param>
+		/// <returns>return true or false</returns>
+		private bool validCloseOrEmptyChars(string stringCharacters) {
```
there are a couple of things you could improve here:
the function name + parameter suggests it will compare characters (in plural) to close or empty chars.
however the function only validates the starting character (position 0).
If you see your 2 functions validOpenChars and validCloseChars, the name looks similar but they do different things.

So my suggestion would be change the function name to something like `isFirstCharCloseOrEmpty` or do something similar than with validOpenChars, including the position

jorgetinoco replied:
The other thing, is that this line is redundant:
return (boolean statement) ? true : false;

you can always `return boolean statement`

jorgetinoco commented on Algorithms/ValidParentheses.cs:
```diff
@@ -22,22 +22,38 @@ public class ValidParentheses
+			return characters.Count == 0 & true;
+		}
+		/// <summary>
+		/// Valid open chars
```
If you are going to include comments, make them more useful than just the name of the function

jorgetinoco commented on Algorithms/ValidParentheses.cs:
```diff
@@ -22,22 +22,38 @@ public class ValidParentheses
+		/// Validate start of string with closing characters and string is empty
+		/// </summary>
+		/// <param name="stringCharacters">string of characters</param>
+		/// <returns>return true or false</returns>
```
when does it returns true or false? what needs to happen to get each?

jorgetinoco commented:
@rogs-lv just checking in if you had time to review the latest comments 😄

rogs-lv committed:
Correct method names and descriptions

rogs-lv commented:
It's ready 👍  💯  😃
